README.md
#Tetris
Tetris game


history.txt
В 1984 году, 35 лет назад, сотрудник вычислительного центра Академии
наук СССР Алексей Пажитнов разработал компьютерную игру-пазл "Тетрис".

Пажитнов не занимался специально разработкой "Тетриса". Простой инженер
из вычислительного центра Академии наук СССР тестировал мощность нового
компьютера — "Электроника 60". Этот тест проводился с помощью
специальной программы, и на дисплее появлялись текстовые символы,
которые заполняли все свободное место. "Отработанные" строки Алексей
удалял. И вот тогда-то инженеру и пришла мысль об головоломке.

Алексею в то время было всего 29 лет, и он, конечно же, увлекался
головоломками Голомба. Он захотел создать нечто подобное в компьютерном
варианте. Правда, элементы в головоломке состояли из 5 кубов, и это
было большим недостатком игры. Потому что компьютеры в то время не
справлялись с обработкой информации. И тогда Алексей решил ограничиться
в своей игре не пятью элементами, как в пентамино (в переводе с
греческого "пента" — пять), а четырьмя (тетра). Отсюда и название игры — "Тетрис".

Первыми с игрой познакомились коллеги Пажитнова, и за пределы
лаборатории она пока не выходила. Но затем по просьбе Пажитнова
16-летний вундеркинд Вадим Герасимов перенес головоломку на IBM PC,
то есть американские персональные компьютеры.
Потом партнеры стали копировать игру на дискеты и распространять уже
за пределами НИИ. Так "Тетрис" стала известной в СССР игрой, в нее
играли теперь и дети, и взрослые.
"Тетрисом" заинтересовались и в США, осознав, что эта головоломка
может быть источником получения большой прибыли. Однако самому автору
игры Алексею Пажитнову эту прибыль довелось увидеть лишь через 10 лет.
Именно столько времени ему пришлось бороться за авторские права на
разработанную им игру.


figures.py
import pygame

pygame.init()
size = WIDTH, HEIGHT = 260, 350
screen = pygame.display.set_mode(size)


def draw_smashboy():
    pygame.draw.rect(screen, pygame.Color('yellow'),
                     (10, 10, 25, 25), 0)
    pygame.draw.rect(screen, pygame.Color('yellow'),
                     (40, 10, 25, 25), 0)
    pygame.draw.rect(screen, pygame.Color('yellow'),
                     (10, 40, 25, 25), 0)
    pygame.draw.rect(screen, pygame.Color('yellow'),
                     (40, 40, 25, 25), 0)


def draw_line():
    pygame.draw.rect(screen, pygame.Color('turquoise'),
                     (10, 100, 25, 25))
    pygame.draw.rect(screen, pygame.Color('turquoise'),
                     (10, 130, 25, 25))
    pygame.draw.rect(screen, pygame.Color('turquoise'),
                     (10, 160, 25, 25))
    pygame.draw.rect(screen, pygame.Color('turquoise'),
                     (10, 190, 25, 25))


def draw_orange_ricky():
    pygame.draw.rect(screen, pygame.Color('orange'),
                     (100, 10, 25, 25))
    pygame.draw.rect(screen, pygame.Color('orange'),
                     (100, 40, 25, 25))
    pygame.draw.rect(screen, pygame.Color('orange'),
                     (100, 70, 25, 25))
    pygame.draw.rect(screen, pygame.Color('orange'),
                     (130, 70, 25, 25))


def draw_zig_zag_from_left():
    pygame.draw.rect(screen, pygame.Color('green'),
                     (100, 120, 25, 25))
    pygame.draw.rect(screen, pygame.Color('green'),
                     (100, 150, 25, 25))
    pygame.draw.rect(screen, pygame.Color('green'),
                     (130, 150, 25, 25))
    pygame.draw.rect(screen, pygame.Color('green'),
                     (130, 180, 25, 25))


def draw_blue_ricky():
    pygame.draw.rect(screen, pygame.Color('blue'),
                     (220, 10, 25, 25))
    pygame.draw.rect(screen, pygame.Color('blue'),
                     (220, 40, 25, 25))
    pygame.draw.rect(screen, pygame.Color('blue'),
                     (220, 70, 25, 25))
    pygame.draw.rect(screen, pygame.Color('blue'),
                     (190, 70, 25, 25))


def draw_zig_zag_from_right():
    pygame.draw.rect(screen, pygame.Color('red'),
                     (40, 250, 25, 25))
    pygame.draw.rect(screen, pygame.Color('red'),
                     (10, 280, 25, 25))
    pygame.draw.rect(screen, pygame.Color('red'),
                     (40, 280, 25, 25))
    pygame.draw.rect(screen, pygame.Color('red'),
                     (10, 310, 25, 25))


def draw_t_like():
    pygame.draw.rect(screen, pygame.Color('purple'),
                     (120, 250, 25, 25))
    pygame.draw.rect(screen, pygame.Color('purple'),
                     (120, 280, 25, 25))
    pygame.draw.rect(screen, pygame.Color('purple'),
                     (120, 310, 25, 25))
    pygame.draw.rect(screen, pygame.Color('purple'),
                     (150, 280, 25, 25))


running = True
draw_smashboy()  # квадрат, в левом верхнем углу
draw_line()  # линия, ниже квадрата
draw_orange_ricky()  # уголок вправо, справа от квадрата
draw_zig_zag_from_left()  # зиг заг слеве направо, справа от линии
draw_blue_ricky()  # уголок влево, справа от уголка вправо
draw_zig_zag_from_right()  # зиг заг справа налево, под линией
draw_t_like()  # т образная фигурка, под зигзагами

while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
    pygame.display.flip()
pygame.quit()


the_game.py
import os
import sys

import pygame

pygame.init()
pygame.display.set_caption('Tetris')
SIZE = SCREEN_WIDTH, SCREEN_HEIGHT = 800, 600
screen = pygame.display.set_mode(SIZE)
clock = pygame.time.Clock()
FPS = 60
pygame.key.set_repeat(200, 70)

horizontal_speed = 40
vertical_speed = 10
figure_size = 40, 40
start_pos = 200, 0

all_sprites = pygame.sprite.Group()
all_figures = pygame.sprite.Group()
all_borders = pygame.sprite.Group()
horizontal_borders = pygame.sprite.Group()
vertical_borders = pygame.sprite.Group()


class Border(pygame.sprite.Sprite):
    # строго вертикальный или строго горизонтальный отрезок
    def __init__(self, x1, y1, x2, y2):
        super().__init__(all_borders)
        if x1 == x2:  # вертикальная стенка
            self.add(vertical_borders)
            self.image = pygame.Surface([1, y2 - y1])
            self.mask = pygame.mask.from_surface(self.image)
            self.rect = pygame.Rect(x1, y1, 1, y2 - y1)

        else:  # горизонтальная стенка
            self.add(horizontal_borders)
            self.image = pygame.Surface([x2 - x1, 1])
            self.mask = pygame.mask.from_surface(self.image)
            self.rect = pygame.Rect(x1, y1, x2 - x1, 1)


class Figure(pygame.sprite.Sprite):
    def __init__(self):
        super().__init__(all_figures)
        self.image = load_image('alone_sqr.png')
        self.rect = self.image.get_rect()
        self.rect.size = figure_size
        self.rect.x, self.rect.y = start_pos

    def update(self, *args, **kwargs) -> None:
        if pygame.sprite.groupcollide(
                all_figures, horizontal_borders, False, False) or \
                pygame.sprite.groupcollide(
                    all_figures, all_sprites, False, False):
            all_sprites.add(self)
            all_figures.empty()
            figure()
        else:
            self.rect = self.rect.move(0, 1)
            if args[0] == 'left' and self.rect.x > 0:
                self.rect.x -= horizontal_speed

            if args[0] == 'right' and self.rect.x < SCREEN_WIDTH // 2 - 40:
                self.rect.x += horizontal_speed

            if args[0] == 'down' and self.rect.y < SCREEN_HEIGHT - 55:
                self.rect.y += vertical_speed

        if pygame.sprite.groupcollide(all_sprites, vertical_borders, False, False):
            pass


Border(0, SCREEN_HEIGHT - 5, SCREEN_WIDTH // 2, SCREEN_HEIGHT - 5)
Border(0, 0, 0, SCREEN_HEIGHT - 5)
Border(SCREEN_WIDTH // 2, 0, SCREEN_WIDTH // 2, SCREEN_HEIGHT - 5)


def terminate():
    pygame.quit()
    sys.exit()


def load_image(name, colorkey=None):
    fullname = os.path.join('data', name)
    if not os.path.isfile(fullname):
        print(f"Файл с изображением '{fullname}' не найден")
        sys.exit()
    image = pygame.image.load(fullname)

    if colorkey is not None:
        image = image.convert()
        if colorkey == -1:
            colorkey = image.get_at((0, 0))
        image.set_colorkey(colorkey)
    else:
        image = image.convert_alpha()
    return image


figure = Figure
figure()

while True:
    screen.fill('white')
    side = None
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            terminate()

        keyboard_keys = pygame.key.get_pressed()

        if keyboard_keys[pygame.K_LEFT]:
            side = 'left'

        elif keyboard_keys[pygame.K_RIGHT]:
            side = 'right'

        if keyboard_keys[pygame.K_DOWN]:
            side = 'down'

    all_sprites.draw(screen)
    all_figures.draw(screen)
    all_borders.draw(screen)
    all_figures.update(side)
    clock.tick(FPS)
    pygame.display.flip()


main.py
import os
import sys
from copy import deepcopy
from random import randrange, choice

import pygame
import pygame_gui

pygame.init()
pygame.display.set_caption('Tetris')
SIZE = SCREEN_WIDTH, SCREEN_HEIGHT = 800, 700
screen = pygame.display.set_mode(SIZE)
clock = pygame.time.Clock()
manager = pygame_gui.UIManager(SIZE)
FPS = 60

SCORE = 0
BEST_SCORE = 0

scale = 30
W, H = 10, 20
TOP = 30
LEFT = 30

figuresPos = [[(-1, 0), (-2, 0), (0, 0), (1, 0)],
              [(0, -1), (-1, -1), (-1, 0), (0, 0)],
              [(-1, 0), (-1, 1), (0, 0), (0, -1)],
              [(0, 0), (-1, 0), (0, 1), (-1, -1)],
              [(0, 0), (0, -1), (0, 1), (-1, -1)],
              [(0, 0), (0, -1), (0, 1), (1, -1)],
              [(0, 0), (0, -1), (0, 1), (-1, 0)]]

MOVE_FIGURE = pygame.USEREVENT + 1
pygame.time.set_timer(MOVE_FIGURE, 1000)


def terminate():
    pygame.quit()
    sys.exit()


def load_image(name, colorkey=None):
    fullname = os.path.join('data', name)
    if not os.path.isfile(fullname):
        print(f"Файл с изображением '{fullname}' не найден")
        sys.exit()
    image = pygame.image.load(fullname)

    if colorkey is not None:
        image = image.convert()
        if colorkey == -1:
            colorkey = image.get_at((0, 0))
        image.set_colorkey(colorkey)
    else:
        image = image.convert_alpha()

    return image


def write_history():
    manager_hist = pygame_gui.UIManager(SIZE)
    fon = pygame.transform.scale(load_image('images/historyMenuBackground.jpg'),
                                 (SCREEN_WIDTH, SCREEN_HEIGHT))
    screen.blit(fon, (0, 0))

    with open('history.txt', 'r', encoding='utf-8') as hist:
        file = [line.strip() for line in hist]
    font = pygame.font.Font(None, 20)

    for i in range(len(file)):
        text = font.render(file[i], True, pygame.Color('white'))
        text_x = SCREEN_WIDTH // 2 - text.get_width() // 2
        text_y = 5
        screen.blit(text, (text_x, text_y + 20 * i))

    buttonWidth = 100
    buttonHeight = 50
    buttonTop = 0
    buttonStartGame = pygame_gui.elements.UIButton(
        relative_rect=pygame.Rect((buttonTop, buttonTop),
                                  (buttonWidth, buttonHeight)),
        text='Back',
        manager=manager_hist)

    while True:
        time_delta = clock.tick(60) / 1000.0
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                terminate()
            if event.type == pygame.USEREVENT:
                if event.user_type == pygame_gui.UI_BUTTON_PRESSED:
                    if event.ui_element == buttonStartGame:
                        start_screen()
            manager_hist.process_events(event)

        manager_hist.update(time_delta)
        pygame.display.flip()
        manager_hist.draw_ui(screen)
        clock.tick(FPS)


def start_screen():
    manager_screen = pygame_gui.UIManager(SIZE)
    line = 'Tetris'

    fon = pygame.transform.scale(load_image('images/startScreenBackground.jpg'),
                                 (SCREEN_WIDTH, SCREEN_HEIGHT))
    screen.blit(fon, (0, 0))

    font = pygame.font.Font(None, 100)
    string_rendered = font.render(line, True, pygame.Color('black'))
    intro_rect = string_rendered.get_rect()
    intro_rect.top = 100
    intro_rect.x = SCREEN_WIDTH // 2 - 100
    screen.blit(string_rendered, intro_rect)

    buttonWidth = 300
    buttonHeight = 50
    buttonTop = 250
    buttonSpace = 20

    buttonStartGame = pygame_gui.elements.UIButton(
        relative_rect=pygame.Rect(((SCREEN_WIDTH - buttonWidth) // 2, buttonTop),
                                  (buttonWidth, buttonHeight)),
        text='Start Game',
        manager=manager_screen)

    buttonSettings = pygame_gui.elements.UIButton(
        relative_rect=pygame.Rect(((SCREEN_WIDTH - buttonWidth)
                                   // 2, buttonTop + buttonHeight + buttonSpace),
                                  (buttonWidth, buttonHeight)),
        text='Settings',
        manager=manager_screen)

    buttonHistory = pygame_gui.elements.UIButton(
        relative_rect=pygame.Rect(((SCREEN_WIDTH - buttonWidth) // 2, buttonTop + 2 * (buttonHeight + buttonSpace)),
                                  (buttonWidth, buttonHeight)),
        text='History',
        manager=manager_screen)

    buttonExit = pygame_gui.elements.UIButton(
        relative_rect=pygame.Rect(((SCREEN_WIDTH - buttonWidth) // 2,
                                   buttonTop + 3 * (buttonHeight + buttonSpace)),
                                  (buttonWidth, buttonHeight)),
        text='Exit',
        manager=manager_screen)

    while True:
        time_delta = clock.tick(60) / 1000.0
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                terminate()

            if event.type == pygame.USEREVENT:
                if event.user_type == pygame_gui.UI_BUTTON_PRESSED:
                    if event.ui_element == buttonStartGame:
                        newGame()

                    if event.ui_element == buttonSettings:
                        print('Settings!')

                    if event.ui_element == buttonHistory:
                        write_history()

                    if event.ui_element == buttonExit:
                        terminate()

            manager_screen.process_events(event)

        manager_screen.update(time_delta)

        pygame.display.flip()
        manager_screen.draw_ui(screen)
        clock.tick(FPS)


def newGame():
    global SCORE, BEST_SCORE
    SCORE = 0
    grid = Grid(W, H, TOP, LEFT)

    screen.fill('black')
    f = Figure(figuresPos[randrange(len(figuresPos))])
    f.draw()

    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                terminate()
            elif event.type == MOVE_FIGURE:
                screen.fill('black')
                if f.move:
                    f.update(grid)
                else:
                    f = Figure(figuresPos[randrange(len(figuresPos))])
                    f.draw()
            keyboard_keys = pygame.key.get_pressed()

            if keyboard_keys[pygame.K_LEFT]:
                side = -1
                f.moveHorizontal(side, grid)

            elif keyboard_keys[pygame.K_RIGHT]:
                side = 1
                f.moveHorizontal(side, grid)

            if keyboard_keys[pygame.K_DOWN]:
                screen.fill('black')
                f.update(grid)

            if keyboard_keys[pygame.K_UP]:
                screen.fill('black')
                f.turn_figure(grid)

        grid.draw()
        grid.score(SCORE, BEST_SCORE)
        clock.tick(FPS)
        pygame.display.flip()


class Figure:
    imagesFigure = [load_image('images/alone_sqr1.png'), load_image('images/alone_sqr2.png'),
                    load_image('images/alone_sqr3.png')]

    def __init__(self, structure):
        self.structure = structure
        self.spitesFigure = pygame.sprite.Group()
        self.move = True
        img = choice(Figure.imagesFigure)

        for pos in self.structure:
            dx, dy = pos
            sqr = pygame.sprite.Sprite(self.spitesFigure)
            sqr.image = img
            sqr.rect = sqr.image.get_rect()
            sqr.rect.x = W // 2 * scale + dx * scale + LEFT
            sqr.rect.y = dy * scale + TOP

    def draw(self):
        self.spitesFigure.draw(screen)

    def turn_figure(self, grid):
        figure = []
        for spr in self.spitesFigure:
            figure.append(spr.rect)
        center = figure[0]
        on_board = 0
        for i in range(4):
            x = figure[i].y - center.y
            y = figure[i].x - center.x
            x1 = center.x - x
            y1 = center.y + y
            if (scale <= x1 + scale <= (W + 1) * scale and
                scale <= x1 <= (W + 1) * scale) and \
                    (scale <= y1 + scale <= (H + 1) * scale and
                     scale <= y1 <= (H + 1) * scale) and not grid.field[y1 // 30 - 1][x1 // 30 - 1]:
                on_board += 1
            else:
                break
        if on_board == 4:
            for i in range(4):
                x = figure[i].y - center.y
                y = figure[i].x - center.x
                figure[i].x = center.x - x
                figure[i].y = center.y + y
        self.spitesFigure.draw(screen)

    def moveHorizontal(self, side, grid):
        for spr in self.spitesFigure:
            x, y = (spr.rect.x - LEFT) // scale, (spr.rect.y - TOP) // scale
            if not (0 <= x + side < W) or grid.field[y][x + side]:
                return

        for spr in self.spitesFigure:
            spr.rect.x += scale * side
        screen.fill('black')
        self.draw()

    def update(self, grid):
        self.collision(grid)
        if self.move:
            for spr in self.spitesFigure:
                spr.rect.y += scale
            self.draw()
        else:
            self.destruct(grid)

    def collision(self, grid):
        global BEST_SCORE
        for spr in self.spitesFigure:
            x = (spr.rect.x - LEFT) // scale
            y = (spr.rect.y - TOP) // scale
            if y + 1 >= H or grid.field[y + 1][x]:
                self.move = False
                if y <= 0:
                    print('You lose!')
                    if BEST_SCORE < SCORE:
                        BEST_SCORE = SCORE
                    newGame()
                return

    def destruct(self, grid):
        for spr in self.spitesFigure:
            x = (spr.rect.x - LEFT) // scale
            y = (spr.rect.y - TOP) // scale
            grid.field[y][x] = 1
            grid.newSprite(spr)
        grid.update()


class Grid:
    def __init__(self, w, h, top, left):
        self.width = w
        self.height = h
        self.top = top
        self.left = left

        self.field = [[0] * w for _ in range(h)]
        self.spritesSqr = pygame.sprite.Group()

    def draw(self):
        for x in range(self.width):
            for y in range(self.height):
                pygame.draw.rect(screen, 'white', ((x * scale + self.left,
                                                    y * scale + self.top),
                                                   (scale, scale)),
                                 width=1)
        self.spritesSqr.draw(screen)

    def update(self):
        global SCORE
        for y in range(self.height):
            if all(self.field[y]):
                self.field = [[0] * self.width] + self.field[:y] + self.field[y + 1:]

                coord = y * scale + TOP
                for spr in self.spritesSqr:
                    if spr.rect.y == coord:
                        self.spritesSqr.remove(spr)
                    elif spr.rect.y < coord:
                        spr.rect.y += scale
                SCORE += 100
        screen.fill('black')
        self.draw()

    def newSprite(self, sprite):
        self.spritesSqr.add(sprite)

    def score(self, score, best_score):
        texts = ["SCORE", score, "BEST SCORE", best_score]
        for i in range(4):
            font = pygame.font.Font(None, 50)
            text = font.render(str(texts[i]), True, pygame.Color('grey'))
            text_x = SCREEN_WIDTH // 2 + 100
            text_y = 200 + i * 50
            screen.blit(text, (text_x, text_y))


start_screen()
